### PR TITLE
topology: pre-processor: Introduce extends/overrides keywords for classes

### PR DESCRIPTION
**Pull request by ranj063**
Introduce a new keyword "Subtreecopy" for extending existing conf nodes
with additional nodes. This feature is useful for extending previous
pipeline class definitions with the addition of one or more widgets
without having to duplicate everything in the new class definition.

For example: Consider a pipeline class definition as below. Note that
only the widgets & routes are shown here.
```
Class.Pipeline.mixout-gain-dai-copier-playback {
	Object.Widget {
		mixout."1" {}
		dai-copier."1" {}
		gain."1" {}
		pipeline."1" {}
	}

	Object.Base {
		!route [
			{
				source mixout.$index.1
				sink	gain.$index.1
			}
		]
	}
}
```
If we want to extend this pipeline with the addition of an eqiir/eqfir,
we can create a Subtreecopy node with type extend as follows:
```
Subtreecopy.mixout-gain-eqiir-eqfir-dai-copier-playback {
	source "Class.Pipeline.mixout-gain-dai-copier-playback"
	target "Class.Pipeline"
	type extend

	Object.Widget {
		eqiir.1 {}
		eqfir.1 {}
	}

	Object.Base {
		!route [
			{
				source gain.$index.1
				sink   eqiir.$index.1
			}
			{
				source eqiir.$index.1
				sink   eqfir.$index.1
			}
		]
	}
}
```
But if we want to modify an existing pipeline class while modifying the
order of widgets and/or inserting new widgets, we should use the type
"override" instead. This allows for adding new widgets to the list of
widgets in the base class definition while also allowing overriding the
routes to allow inserting the new widgets and reordering the widgets in
the base class. For example, if we want to add a drc widget between the
gain and the eqiir modules in the above class, we can do the following:
```
Subtreecopy.mixout-efx-dai-copier-playback {
	source "Class.Pipeline.mixout-gain-eqiir-eqfir-dai-copier-playback"
	target "Class.Pipeline"
	type override

	Object.Widget {
		drc.1 {}
	}

	Object.Base {
		!route [
			{
				source mixout.$index.1
				sink	gain.$index.1
			}
			{
				source gain.$index.1
				sink	drc.$index.1
			}
			{
				source	drc.$index.1
				sink	eqiir.$index.1
			}
			{
				source	eqiir.$index.1
				sink	eqfir.$index.1
			}
		]
	}
}
```
Note that the routes array contains all the routes in the new subtreecopy definition.

This subtreecopy feature can also be used to copy normal config blocks
without having the need to duplicate information.

For example, if all the widgets in a pipeline support the same audio
formats, we can define it as follows:
```
Class.Pipeline.dai-copier-eqiir-module-copier-capture {
	formats {
		num_input_audio_formats 2
		num_output_audio_formats 2

		Object.Base.input_audio_format [
			{
				in_bit_depth		32
				in_valid_bit_depth	32
			}
			{
				in_channels		4
				in_bit_depth		32
				in_valid_bit_depth	32
				in_ch_cfg	$CHANNEL_CONFIG_3_POINT_1
				in_ch_map	$CHANNEL_MAP_3_POINT_1
			}
		]

		Object.Base.output_audio_format [
			{
				out_bit_depth		32
				out_valid_bit_depth	32
			}
			{
				out_channels		4
				out_bit_depth		32
				out_valid_bit_depth	32
				out_ch_cfg	$CHANNEL_CONFIG_3_POINT_1
				out_ch_map	$CHANNEL_MAP_3_POINT_1
			}
		]
	}

	Object.Widget {
		dai-copier."1" {
			type dai_out
			num_output_pins 1

			Subtreecopy.formats {
				source "Class.Pipeline.dai-copier-eqiir-module-copier-capture.formats"
				type override
			}
		}

		eqiir."1" {
			Subtreecopy.formats {
				source "Class.Pipeline.dai-copier-eqiir-module-copier-capture.formats"
				type override
			}

			Object.Control.bytes."1" {
				IncludeByKey.DMIC0_DAI_EQIIR {
					"passthrough"		"include/components/eqiir/passthrough.conf"
					"highpass_40hz_0db"	"include/components/eqiir/highpass_40hz_0db_48khz.conf"
					"highpass_40hz_20db"	"include/components/eqiir/highpass_40hz_20db_48khz.conf"
				}
			}
		}

		module-copier."2" {
			Subtreecopy.formats {
				source "Class.Pipeline.dai-copier-eqiir-module-copier-capture.formats"
				type override
			}
		}

		pipeline."1" {
			priority	0
			lp_mode		0
		}
	}
}
```
The subtreecopy node in each widget ensures that the pipeline formats
are applied to all widgets without the need for duplicating them for
each widget.

Link: https://github.com/thesofproject/sof/issues/9082

### CODE REVIEW

**Comment by ranj063** [comment]
@plbossart @lrgirdwo @jsarha FYI

**Comment by plbossart** [comment]
Thanks @ranj063 
This feature would simplify the life of maintainers and 3rd parties, we have way too much duplication of entire pipelines when we add an additional processing element. The use of tried-and-tested templates that can be simply extended in a nested manner is probably the best thing since we moved over from M4 to topology2!

**Comment by ranj063** [comment]
> Could we just work with whole configuration blocks? This feature may be useful for all config parts. Like:
> 
> ```
> SubtreeCopy.AnUniqueId {
>   type override
>   source "Class.Pipeline.mixout-gain-eqiir-eqfir-dai-copier-playback"
>   target "Class.Pipeline.mixout-efx-dai-copier-playback"
>   tree {
>     Object.Widget {
> 	drc.1 {}
>     }
> 
>     Object.Base {
>          !route [
>             ...
>          ]
>     } 
>   }
> }
> ```

@perexg since we'd like to keep the new sub class definitions in separate conf files, the uniqueid might be problematic. But maybe if I make it an array like below, it might be OK.
Subtreecopy [
     {
      ....
     }
] 

Also, can I retain both the extends/overrides type for the subtreecopy to achieve both pipeline extensions and pipeline widget insertion+reordering for Pipeline classes. For other configs, this could work as a generic config block copy.

**Comment by perexg** [comment]
> @perexg since we'd like to keep the new sub class definitions in separate conf files, the uniqueid might be problematic. But maybe if I make it an array like below, it might be OK. Subtreecopy [ { .... } ]

Unique ID may be like "Pipeline-mixout-efx-dai-copier-playback" so I don't see a problem here. Using [] (arrays) will assign IDs from the index range (0 ... ) automatically.

> Also, can I retain both the extends/overrides type for the subtreecopy to achieve both pipeline extensions and pipeline widget insertion+reordering for Pipeline classes. For other configs, this could work as a generic config block copy.

The merge type for the new tree update may be specified using a field - I proposed `type` in my example.

**Comment by plbossart** [comment]
Not sure I follow what the uniqueID is, it looks redundant with the target in the suggested example?

````
SubtreeCopy.Pipeline-mixout-efx-dai-copier-playback {
  type override
  source "Class.Pipeline.mixout-gain-eqiir-eqfir-dai-copier-playback"
  target "Class.Pipeline.mixout-efx-dai-copier-playback"
````

**Comment by ranj063** [comment]
@perexg I have modifed the PR to define the Subtreecopy keyword now. Please let me know what you think of this. Thanks!

**Comment by perexg** [comment]
> Not sure I follow what the uniqueID is, it looks redundant with the target in the suggested example?
> 
> ```
> SubtreeCopy.Pipeline-mixout-efx-dai-copier-playback {
>   type override
>   source "Class.Pipeline.mixout-gain-eqiir-eqfir-dai-copier-playback"
>   target "Class.Pipeline.mixout-efx-dai-copier-playback"
> ```

It's just up to the user to create a unique ID scheme. You can compose any string. If you use one key for multiple configs, the last value is used:

```
a.a "1"
a.a "2"
```

The result is that `a.a = "2"` when the configuration tree is parsed.

File: topology/pre-processor.c [hunk 1]
```diff
@@ -921,6 +921,260 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+		}
+
+		/* handle all other subtree copies */
+		ret = snd_config_copy(&tmp, source_tree);
+		if (ret < 0) {
+			SNDERR("failed to copy source tree for subtreecopy %s\n", id);
+			return ret;
+		}
+
+		ret = snd_config_merge(tmp, n, override);
```
Comment (perexg) [inline topology/pre-processor.c:1128]: Why is this required? Only `source_tree` should be merged to the `target_tree` here.

File: topology/pre-processor.c [hunk 2]
```diff
@@ -921,6 +921,260 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+}
+
+static int pre_process_subtree_copy(struct tplg_pre_processor *tplg_pp, snd_config_t *curr,
+				    snd_config_t *top)
+{
+	snd_config_iterator_t i, next;
+	snd_config_t *subtrees;
+	int ret;
+
+	ret = snd_config_search(curr, "Subtreecopy", &subtrees);
```
Comment (perexg) [inline topology/pre-processor.c:1016]: We have already special blocks like `CombineArrays`, thus using upper cases in the middle is fine. I would prefer `SubtreeCopy` or `SubTreeCopy` here or even `TreeCopy` or `ConfigCopy` (may be a good alternative).

File: topology/pre-processor.c [hunk 3]
```diff
@@ -921,6 +921,260 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+		ret = snd_config_get_id(n, &id);
+		if (ret < 0) {
+			SNDERR("Failed to get ID for subtree copy\n");
+			return ret;
+		}
+
+		/* get the type of copy ex: override/extend */
+		ret = snd_config_search(n, "type", &type_cfg);
+		if (ret < 0) {
+			SNDERR("failed to get type for subtree %s\n", id);
```
Comment (perexg) [inline topology/pre-processor.c:1037]: O would keep this as optional with a default (`override` ?).

File: topology/pre-processor.c (hunk 1 above)
Comment (perexg) [reply topology/pre-processor.c:1128]: Probably a missing lookup for the `tree` field before.

**Comment by perexg** [comment]
Ok, I would remove all class.pipeline specific code. It is not required IMHO. See:

```
Class.Pipeline.mixout-efx-dai-copier-playback {
       Object.Base {
         !route [
            ...
         ]
    }
}

SubtreeCopy.add-to-mixout-efx {
  type override
  source "Class.Pipeline.mixout-gain-eqiir-eqfir-dai-copier-playback.Object.Widget"
  target "Class.Pipeline.mixout-efx-dai-copier-playback.Object.Widget"
}
```

File: topology/pre-processor.c [hunk 4]
```diff
@@ -921,6 +921,260 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+		if (ret >= 0) {
+			ret = snd_config_get_string(target_cfg, &target_id);
+			if (ret < 0) {
+				SNDERR("Invalid source node id for subtree %s\n", id);
+				return ret;
+			}
+
+			ret = snd_config_search(top, target_id, &target_tree);
+			if (ret < 0) {
+				SNDERR("failed to get target tree %s\n", target_id);
```
Comment (perexg) [inline topology/pre-processor.c:1071]: You should create an empty compound for the given target path (multi id) here.

Comment (ranj063) [reply topology/pre-processor.c:982]: @perexg I think I have addressed all your comments except this one. Im not clear why I need to create an empty compound here. I'm creating a new conf block with the source tree and the subtree merged with/without override and then I merge the merged tree to the target tree.

Comment (perexg) [reply topology/pre-processor.c:982]: But the target subtree in the top tree may not exist yet and it's not an error. In this case, an empty tree with target_id path should be created to work with it in the merge operation. Probably the easiest implementation is to use `snd_config_load_string` with "<target_id> {}" string as source and merge the result to `top` and do new `snd_config_search` for `target_id` / `target_tree`.

**Comment by perexg** [comment]
BTW: I like your nested examples. They show the power for the subtree copying.

File: topology/pre-processor.c [hunk 5]
```diff
@@ -921,6 +921,160 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+		}
+
+		ret = snd_config_search(top, source_id, &source_tree);
+		if (ret < 0) {
+			SNDERR("failed to get source tree %s\n", source_id);
+			return ret;
+		}
+
+		/* delete the source/target/type nodes */
+		ret = snd_config_delete(source_cfg);
```
Comment (perexg) [inline topology/pre-processor.c:1003]: `snd_config_delete(subtrees)` does this job for the whole processed `SubTreeCopy` block.  I think that this step is extra.

File: topology/pre-processor.c [hunk 6]
```diff
@@ -921,6 +921,160 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+		if (ret < 0)
+			return ret;
+
+		if (type) {
+			ret = snd_config_delete(type_cfg);
```
Comment (perexg) [inline topology/pre-processor.c:1008]: I would recheck all error paths and the resource deallocation in this function. There are many resource leaks IMHO.

**Review by perexg — CHANGES_REQUESTED** [review]

**Comment by ranj063** [comment]
> The `tree` field removal means data mixup (what if "source" id is a valid id in the copied sub-tree for example?). My proposal was to have SubTreeCopy configuration separated.

@perexg Sorry for the delay, it was a long weekend in the US. I removed the target/source/type fields from the SUbTreeCOpy node explicitly so that if those fields are valid in the subtree being copied, they will be retained.

File: topology/pre-processor.c (hunk 4 above)
Comment (ranj063) [reply topology/pre-processor.c:1006]: ack, fixed now

File: topology/pre-processor.c (hunk 6 above)
Comment (ranj063) [reply topology/pre-processor.c:1035]: The only 2 resources allocated are for the temporary target ID and the temporary node while merging the source tree. I think I have freed both of them in the error paths.

**Comment by perexg** [comment]
> > The `tree` field removal means data mixup (what if "source" id is a valid id in the copied sub-tree for example?). My proposal was to have SubTreeCopy configuration separated.
> 
> @perexg Sorry for the delay, it was a long weekend in the US. I removed the target/source/type fields from the SUbTreeCOpy node explicitly so that if those fields are valid in the subtree being copied, they will be retained.

Yes, but you cannot override them later then. In my eyes, it would be better to keep the source subtree completely separated (although it requires a new tab level).

**Comment by ranj063** [comment]
> Yes, but you cannot override them later then. In my eyes, it would be better to keep the source subtree completely separated (although it requires a new tab level).

@perexg I'm not sure I'm following your comment. Could you please elaborate what you're thinking here?

**Comment by perexg** [comment]
Do not mix SubTreeCopy configuration with the added/modified tree.

```
SubTreeCopy.### {
  type ###
  source ###
  target ###
  tree {
    ### changes here ###
  }
}
```

Vs. your current implementation:

```
SubTreeCopy.### {
  type ###
  source ###
  target ###
  ### changes here ###
}
```

**Comment by ranj063** [comment]
> Do not mix SubTreeCopy configuration with the added/modified tree.
> 
> ```
> SubTreeCopy.### {
>   type ###
>   source ###
>   target ###
>   tree {
>     ### changes here ###
>   }
> }
> ```
> 
> Vs. your current implementation:
> 
> ```
> SubTreeCopy.### {
>   type ###
>   source ###
>   target ###
>   ### changes here ###
> }
> ```

Thanks @perexg. Updated now

**Comment by ranj063** [comment]
@perexg can I please bother you to have another look at this PR? I've addressed all your comments.

File: topology/pre-processor.c [hunk 7]
```diff
@@ -921,6 +921,185 @@ static int pre_process_arrays(struct tplg_pre_processor *tplg_pp, snd_config_t *
...
+		ret = snd_config_search(top, source_id, &source_tree);
+		if (ret < 0) {
+			SNDERR("failed to get source tree %s\n", source_id);
+			return ret;
+		}
+
+		/* get the subtree to be merged. create an empty tree if it dosen't exist */
+		ret = snd_config_search(n, "tree", &subtree_cfg);
+		if (ret < 0) {
+			char *s;
```
Comment (perexg) [inline topology/pre-processor.c:1029]: It's too overkill. Just assign `subtree_cfg = NULL` here. The function `snd_config_merge` does nothing when `src == NULL`.

**Review by perexg — CHANGES_REQUESTED** [review]
Please, do the last little change. Thank you.

File: topology/pre-processor.c (hunk 7 above)
Comment (ranj063) [reply topology/pre-processor.c]: @perexg sorry for the delay. I was out on vacation. Fixed now.

Comment (ranj063) [reply topology/pre-processor.c]: @perexg Could I please request to have a look at the PR again?

**Comment by ranj063** [comment]
@perexg Could you please have a look at this again? This will help us update our SOF topologies. Thanks!

**Pull request closed without merging**